### PR TITLE
[consensus][restart] commit up to highest ledger info after rebuild

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

During restart, if we have LI(S) as ancestor of LI(C), we'll start from LI(S).

This PR enforces us to commit to LI(C) so that  we provide guarantee LI(C) == LI(S) after start consensus, this also simplifies the reasoning of restart + reconfiguration.

We can probably further improve it by having commit triggered by `insert_quorum_cert`.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Existing tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 1]
```diff
@@ -148,6 +151,62 @@ impl<T: Payload> BlockStore<T> {
         tree
     }
 
+    /// Commit the given block id with the proof, returns the path from current root or error
+    pub async fn commit(
+        &self,
+        block_id_to_commit: HashValue,
+        finality_proof: LedgerInfoWithSignatures,
```
Comment (aching) [inline consensus/src/chained_bft/block_storage/block_store.rs]: This PR didn't introduce this, but rather continues the pattern of passing both the block_id_to_commit and the finality_proof.  Can't we just derive the block_id_to_commit from finality_proof.ledger_info().consensus_block_id() and not pass in block_id_to_commit?

File: consensus/src/chained_bft/event_processor.rs [hunk 2]
```diff
@@ -732,64 +729,25 @@ impl<T: Payload> EventProcessor<T> {
...
-            .commit(payload_and_output_list, finality_proof)
+            .commit(block_id_to_commit, finality_proof)
             .await
         {
-            // We assume that state computer cannot enter an inconsistent state that might
-            // violate safety of the protocol. Specifically, an executor service is going to panic
-            // if it fails to persist the commit requests, which would crash the whole process
-            // including consensus.
-            error!(
-                "Failed to persist commit, mempool will not be notified: {:?}",
```
Comment (aching) [inline consensus/src/chained_bft/event_processor.rs]: If the commit fails, why shouldn't we panic!?

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 3]
```diff
@@ -172,6 +231,16 @@ impl<T: Payload> BlockStore<T> {
             error!("fail to delete block: {:?}", e);
         }
         *self.inner.write().unwrap() = tree;
+        // Commit any blocks we can, this helps if we crash before state computer
+        // successfully persist the commits
```
Comment (aching) [inline consensus/src/chained_bft/block_storage/block_store.rs]: Can you please add more detail to what this helps with?  I'm a bit confused by the comment.

File: consensus/src/chained_bft/event_processor.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/chained_bft/event_processor.rs]: see the comment we had above, it's assumed the whole process would crash

Comment (aching) [reply consensus/src/chained_bft/event_processor.rs]: Why not also panic here then?  If somehow the other behavior changed, we would want to panic here right?

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 3 above)
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: it's also described in the pr, if we crash during commit B_n, after restart we'll start from B_i <- B_i+1 ... <- B_n and move on to process proposal/vote etc pretending we never commit B_n.

This helps to commit to B_n first and start processing proposal/vote.

Comment (aching) [reply consensus/src/chained_bft/block_storage/block_store.rs]: Thanks for the explanation.

Can you please update the comment to reflect this, it's not very intuitive.

**Pull request closed without merging**